Always pass a span to String/StringImpl's createUninitialized()

cdumez opened the pull request:
#### 98dc165d6ab86086c87d419d5fbd8d198826e76d
<pre>
Always pass a span to String/StringImpl&apos;s createUninitialized()
<a href="https://bugs.webkit.org/show_bug.cgi?id=282175">https://bugs.webkit.org/show_bug.cgi?id=282175</a>

Reviewed by Geoffrey Garen.

* Source/JavaScriptCore/runtime/FuzzerPredictions.cpp:
(JSC::readFileIntoString):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/cf/CFURLExtras.cpp:
(WTF::bytesAsString):
* Source/WTF/wtf/persistence/PersistentCoders.cpp:
(WTF::Persistence::decodeStringText):
* Source/WTF/wtf/text/AtomStringImpl.cpp:
(WTF::HashedUTF8CharactersTranslator::translate):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::convertToUppercaseWithoutLocaleStartingAtFailingIndex8Bit):
(WTF::StringImpl::convertToUppercaseWithoutLocaleUpconvert):
(WTF::StringImpl::convertToLowercaseWithLocale):
(WTF::StringImpl::convertToUppercaseWithLocale):
(WTF::StringImpl::replace):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/StringView.cpp:
(WTF::convertASCIICase):
(WTF::normalizedNFC):
(WTF::makeStringBySimplifyingNewLinesSlowCase):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::make8Bit):
(WTF::String::convertTo16Bit):
* Source/WTF/wtf/text/WTFString.h:
* Source/WTF/wtf/win/LanguageWin.cpp:
(WTF::localeInfo):
* Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
(PAL::TextCodecLatin1::decode):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readString):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::reversed):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::secureText):
* Source/WebKit/Platform/IPC/DaemonCoders.h:
(WebKit::Daemon::Coder&lt;WTF::String&gt;::decodeStringText):
* Source/WebKit/Platform/IPC/DaemonDecoder.cpp:
(WebKit::Daemon::Decoder::decodeFixedLengthData):
(WebKit::Daemon::Decoder::decodeFixedLengthReference):
* Source/WebKit/Platform/IPC/DaemonDecoder.h:
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeSpan):

Canonical link: <a href="https://commits.webkit.org/285785@main">https://commits.webkit.org/285785@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca204ec099ee332e47a2cfa379ae8373c0bcd05b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73763 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53192 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26574 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78085 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25001 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75880 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/977 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58033 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16402 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76830 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48167 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63476 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38432 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44955 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20964 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23334 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66900 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66527 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21311 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79652 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73021 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1080 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/538 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66375 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1222 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63486 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65654 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9513 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7699 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94802 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11381 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1044 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3794 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20849 "Found 2 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1073 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1060 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1079 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
```diff
@@ -229,9 +229,9 @@ JSC_DEFINE_HOST_FUNCTION(uint8ArrayPrototypeToHex, (JSGlobalObject* globalObject
+    std::span<LChar> buffer;
     auto result = StringImpl::createUninitialized(length * 2, buffer);
-    LChar* bufferEnd = buffer + length * 2;
+    LChar* bufferEnd = buffer.data() + length * 2;
```
Would be nice if this was something more like `buffer.end()`, but I guess that’s an iterator not a pointer, and we don’t really have a pointer version of it.

cdumez replied:
I think we can do `std::to_address(buffer.end())`

darinadler replied:
Nice: That's an improvement since we can't get the math wrong.

cdumez replied:
Making this change in https://github.com/WebKit/WebKit/pull/35865

darinadler commented on Source/JavaScriptCore/runtime/StringConstructor.cpp:
```diff
@@ -89,15 +89,15 @@ JSC_DEFINE_HOST_FUNCTION(stringFromCharCode, (JSGlobalObject* globalObject, Call
+            std::span<UChar> buf16Bit;
             auto impl16Bit = StringImpl::createUninitialized(length, buf16Bit);
-            StringImpl::copyCharacters(buf16Bit, { buf8Bit, i });
+            StringImpl::copyCharacters(buf16Bit.data(), buf8Bit.first(i));
```
Seems like we would want a version of `copyCharacters` where the from is also a `span`, to check overrun on the reading side too, assuming the performance cost is not prohibitive. This applies to many other calls of `copyCharacters` below in this patch, too.

cdumez replied:
Yes, I noticed `copyCharacters` but opted to not to that larger refactoring in this patch. I agree though that copyCharacters() would ideally take 2 spans instead of 1.

darinadler commented on Source/WTF/wtf/persistence/PersistentCoders.cpp:
```diff
@@ -114,9 +114,9 @@ static inline std::optional<String> decodeStringText(Decoder& decoder, uint32_t
+    std::span<CharacterType> buffer;
     String string = String::createUninitialized(length, buffer);
-    if (!decoder.decodeFixedLengthData({ reinterpret_cast<uint8_t*>(buffer), length * sizeof(CharacterType) }))
+    if (!decoder.decodeFixedLengthData(spanReinterpretCast<uint8_t>(buffer)))
```
Isn’t this a place where you could use the new cast you came up with, which casts a greater than byte span into a span of the bytes? Or is that only for casting to `span<const uint8_t>`?

cdumez replied:
> casts a greater than byte span into a span of the bytes

That's what it is but I'm not familiar with the new casting function you're referring to. I'm familiar with `byteCast` but I don't think that's it.

darinadler replied:
Yes, not `byteCast`. Let me find the function I’m thinking of. Someone added it recently, and I thought it was you, but I think if it was, you would remember it.

cdumez replied:
I recently added `asByteSpan()` & `asMutableByteSpan()` but this is not a span -> span conversion. This gives you a byte span from an object (the span size ends up being `sizeof(ObjectType)`.

Here, the code I'm replacing very much looks like the implementation of spanReinterpretCast IMO.

darinadler replied:
OK. What I’m looking for is one that works for the safe kind of `reinterpret_cast`, like casting to a byte, but not other less-safe ones.

cdumez replied:
I think I should make `asByteSpan` / `asMutableSpan` work with span inputs probably. Seems like it would do what you expected here.

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/35832

darinadler commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -513,16 +497,16 @@ Ref<StringImpl> StringImpl::convertToUppercaseWithoutLocaleStartingAtFailingInde
 
-    LChar* dest = data8;
-
+    auto destination = data8;
```
Seems like we can just use `data8` directly, since we’re indexing, not bumping a pointer.

cdumez replied:
Addressing in https://github.com/WebKit/WebKit/pull/35827

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Always pass a span to String/StringImpl's createUninitialized()
https://bugs.webkit.org/show_bug.cgi?id=282175

Reviewed by Geoffrey Garen.

* Source/JavaScriptCore/runtime/FuzzerPredictions.cpp:
(JSC::readFileIntoString):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/StdLibExtras.h:
* Source/WTF/wtf/cf/CFURLExtras.cpp:
(WTF::bytesAsString):
* Source/WTF/wtf/persistence/PersistentCoders.cpp:
(WTF::Persistence::decodeStringText):
* Source/WTF/wtf/text/AtomStringImpl.cpp:
(WTF::HashedUTF8CharactersTranslator::translate):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::convertToUppercaseWithoutLocaleStartingAtFailingIndex8Bit):
(WTF::StringImpl::convertToUppercaseWithoutLocaleUpconvert):
(WTF::StringImpl::convertToLowercaseWithLocale):
(WTF::StringImpl::convertToUppercaseWithLocale):
(WTF::StringImpl::replace):
* Source/WTF/wtf/text/StringImpl.h:
* Source/WTF/wtf/text/StringView.cpp:
(WTF::convertASCIICase):
(WTF::normalizedNFC):
(WTF::makeStringBySimplifyingNewLinesSlowCase):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::make8Bit):
(WTF::String::convertTo16Bit):
* Source/WTF/wtf/text/WTFString.h:
* Source/WTF/wtf/win/LanguageWin.cpp:
(WTF::localeInfo):
* Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
(PAL::TextCodecLatin1::decode):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneDeserializer::readString):
* Source/WebCore/rendering/RenderListMarker.cpp:
(WebCore::reversed):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::secureText):
* Source/WebKit/Platform/IPC/DaemonCoders.h:
(WebKit::Daemon::Coder<WTF::String>::decodeStringText):
* Source/WebKit/Platform/IPC/DaemonDecoder.cpp:
(WebKit::Daemon::Decoder::decodeFixedLengthData):
(WebKit::Daemon::Decoder::decodeFixedLengthReference):
* Source/WebKit/Platform/IPC/DaemonDecoder.h:
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::decodeSpan):

Canonical link: https://commits.webkit.org/285785@main
